### DISCUSSION TITLE
v1.51.0

### DISCUSSION THREAD

**Discussion by alextran1502** [immich-app/immich#2025] [Announcements]
# v1.51.0

## Breaking Changes 🚨

This release added a new container to Immich's ecosystem, namely [Typesense](https://typesense.org/). Please update your `docker-compose.yml` file to include the new container and `.env` file to have the new environment variable, which is `TYPESENSE_API_KEY=some-random-text`. 

Please make sure to have your server and mobile app on the same version so that the app can work correctly.

Typesense container is required for now, but we plan to make it optional with the fallback of using Postgres full-text search for the searching mechanism. 

```yaml
  typesense:
    container_name: immich_typesense
    image: typesense/typesense:0.24.0
    environment:
      - TYPESENSE_API_KEY=${TYPESENSE_API_KEY}
      - TYPESENSE_DATA_DIR=/data
    logging:
      driver: none
    volumes:
      - tsdata:/data

[...]

volumes:
  pgdata:
  model-cache:
  tsdata:
```
## Highlighted

This is one of the more extended release breaks (a total of 53 pull requests were merged) that we've had for a while due to some large items that we've been working on, such as:

### Integrating Typesense and CLIP for a better search system

As mentioned in the previous releases, we want a better search system. As Immich aims to be a fully featured Google Photos alternative, we want to provide you with the best search experience possible. We've been working on integrating [Typesense](https://typesense.org/) and [CLIP](https://www.sbert.net/examples/applications/image-search/README.html) into Immich to provide you with a top-notch searching experience for your larger libraries. 

We chose Typesense for its open-source nature, attractive features, and self-hosting capability.

We chose CLIP for its ability to search for images based on their content rather than their metadata. So you can do something like this:

![SCR-20230320-fyi-2](https://user-images.githubusercontent.com/27055614/226406822-4e3043d0-4992-4204-9690-c3fbf4330a67.jpeg)

![SCR-20230320-fz6](https://user-images.githubusercontent.com/27055614/226406866-f141d69c-6462-431f-ab59-b331d26f0319.jpeg)

The CLIP embedding data is a new column in the database. To get that info for your previously uploaded assets, please consider running the ENCODE CLIP job in the admin’s job panel

<img width="854" alt="image" src="https://user-images.githubusercontent.com/27055614/226407082-985de6d9-1dcc-454d-9a37-2ad44e2cb149.png">

We will continue to iterate and improve the search feature in the subsequent releases. We would appreciate any feedback about the initial implementation - what works well and what areas could be improved.

Note: The mobile app does not incorporate this feature yet. It is only applicable to the web in this release. 

### Allow the mobile app to work offline without an internet connection or when the server is down.

Previously, when the server was down, or your phone didn't have an internet connection, tapping on anything in the mobile app would send you back to the login screen. This is because the mobile app couldn’t retrieve any data from the server. We've added a new feature that allows the mobile app to work offline, so you can still view your photos and videos cached or stored on your device even when the server is down or you don't have an internet connection.

In addition, when viewing photos and videos residing on your device, these are no longer loaded from the server but directly from the device. This reduces your mobile data usage and increases the loading speed of thumbnails while scrolling the timeline.

This has been a more significant feature that required a lot of refactoring from @fyfrey over many months-long processes. We're pleased to be able to release this feature to you finally.

As in other releases, we've also added many smaller features and bug fixes, so please check out the full changelog below.

## Thank you

I want to thank everyone who contributed to this release, especially @fyfrey for the huge amount of work he put into the offline mode feature, @jrasm91 and @zackpollard for the refactoring of the server and the integration of the new search system, @martyfuhry for improving the UI/UX of the mobile app, @michelheusschen for improving the web code base and consistency, @bo0tzz for the never-ending stream of CI/CD improvement and answering questions on Discord, and all the new contributors who spared some of their free time to contribute to the project.

This project is for the community, and from the community, I am honored to be a part of it.

----
As always, please consider supporting the project.

🎉 Cheer! 🎉

## Support

<p align="center">
<img src="https://media.giphy.com/media/LStqgGESXW8XnuCv5y/giphy.gif" width="250" title="Loading ~4000 images/videos"> 
</p>

If you find the project helpful and it helps you in some ways, you can support the project [one time](https://github.com/sponsors/alextran1502?frequency=one-time&sponsor=alextran1502) or [monthly](https://github.com/sponsors/alextran1502) from GitHub Sponsors

It is a great way to let me know that you want me to continue developing and working on this project for years to come.

### Server
* feat(server): added additional storage template preset by @Chipwingg in https://github.com/immich-app/immich/pull/1903
* chore(server): bump API version post release by @zackpollard in https://github.com/immich-app/immich/pull/1909
* fix(server): storage template unit test by @michelheusschen in https://github.com/immich-app/immich/pull/1906
* feat(server): Machine learning's image optimisations by @ollywelch in https://github.com/immich-app/immich/pull/1908
* fix(server): exif extraction swapped params by @michelheusschen in https://github.com/immich-app/immich/pull/1914
* feat(server)!: search via TypeSense by @jrasm91 in https://github.com/immich-app/immich/pull/1778
* chore(server): disable TypeSense logging by @alextran1502 in https://github.com/immich-app/immich/pull/1925
* fix(server): duration string parsing by @jrasm91 in https://github.com/immich-app/immich/pull/1923
* feat(server): improve thumbnail relation and updating by @michelheusschen in https://github.com/immich-app/immich/pull/1897
* feat(web,server): explore by @jrasm91 in https://github.com/immich-app/immich/pull/1926
* fix(server): Increase authentication cookie max-age by @bo0tzz in https://github.com/immich-app/immich/pull/1971
* fix(server): update album updatedAt on assets/users removed/added by @fyfrey in https://github.com/immich-app/immich/pull/1977
* fix(server): remove encoded video file on asset delete by @raisinbear in https://github.com/immich-app/immich/pull/1980
* fix(server): timezone mismatch in server tests by @brighteyed in https://github.com/immich-app/immich/pull/1918
* feat(server): CLIP search integration by @alextran1502 in https://github.com/immich-app/immich/pull/1939
* feat(server): Fallback to text search if machine-learning is disabled by @bo0tzz in https://github.com/immich-app/immich/pull/2015
* fix(server): CLIP search return empty result by @alextran1502 in https://github.com/immich-app/immich/pull/2018
* feat(sever): machine-learning image optimisations by @ollywelch in https://github.com/immich-app/immich/pull/1916

### Mobile App
* feat(mobile): Enter server first for login by @martyfuhry in https://github.com/immich-app/immich/pull/1952
* feature(mobile): sync assets, albums & users to local database on device by @fyfrey in https://github.com/immich-app/immich/pull/1759
* fix(mobile): Fixes hero animation on main timeline by @martyfuhry in https://github.com/immich-app/immich/pull/1924
* fix(mobile): iOS backup debug info tile only shown in iOS by @martyfuhry in https://github.com/immich-app/immich/pull/1951
* feat(mobile): Transparent bottom Android navigation bar by @martyfuhry in https://github.com/immich-app/immich/pull/1953
* fix(mobile): do not crash on malformed asset duration by @fyfrey in https://github.com/immich-app/immich/pull/1921
* fix(mobile): Gallery viewer fullscreen edge case by @martyfuhry in https://github.com/immich-app/immich/pull/1959
* fix(mobile): Sorted shared album and added share user doesn't reflect change in album view by @alextran1502 in https://github.com/immich-app/immich/pull/1955
* fix(mobile): add isar source code as a git submodule for F-Droid build by @fyfrey in https://github.com/immich-app/immich/pull/1985
* fix(mobile): Fix iOS login layout issue by @martyfuhry in https://github.com/immich-app/immich/pull/2004
* feature(mobile): allow app to be used offline by @fyfrey in https://github.com/immich-app/immich/pull/1932
* chore(mobile): remove integration test temporarily by @alextran1502 in https://github.com/immich-app/immich/pull/2008
* feature(mobile): no longer wait for background backup in settings by @fyfrey in https://github.com/immich-app/immich/pull/1984
* feat(mobile): Share album name and adaptive shared album display by @martyfuhry in https://github.com/immich-app/immich/pull/2017
* feat(mobile): persist album sort order by @truedem0n in https://github.com/immich-app/immich/pull/1997
* fix(mobile): proper syncing with Recents album on iOS

### Web
* fix(web): restore album drag and drop upload by @michelheusschen in https://github.com/immich-app/immich/pull/1933
* fix(web): nested layout navigation issue by @alextran1502 in https://github.com/immich-app/immich/pull/1936
* fix(web): admin pages layout issue by @michelheusschen in https://github.com/immich-app/immich/pull/1943
* feat(web): require page load to export title by @michelheusschen in https://github.com/immich-app/immich/pull/1956
* feat(web): improve search bar + add to search page by @michelheusschen in https://github.com/immich-app/immich/pull/1957
* feat(web): Make scaling of albums overview more responsive by @matthinc in https://github.com/immich-app/immich/pull/1981
* fix(web): short layout retention after navigation by @michelheusschen in https://github.com/immich-app/immich/pull/1994
* feat(web): improve /auth pages by @michelheusschen in https://github.com/immich-app/immich/pull/1969
* refactor(web): common layout for user pages by @michelheusschen in https://github.com/immich-app/immich/pull/1995
* feat(web): bundle and 'sveltify' leaflet by @michelheusschen in https://github.com/immich-app/immich/pull/1998
* fix(web): storage template extension display by @etnoy in https://github.com/immich-app/immich/pull/2002

### Others
* chore(docs): add missing space in version_announcement_overlay_text_2 for ru-RU locale by @brighteyed in https://github.com/immich-app/immich/pull/1919
* chore(deps): bump docker/setup-buildx-action from 2.4.1 to 2.5.0 by @dependabot in https://github.com/immich-app/immich/pull/1976
* chore(docs): added note on scope of redis optional parameters in example.env by @DragosRotaru in https://github.com/immich-app/immich/pull/1974
* chore(docs): Document how photo dates are determined by @twitsforbrains in https://github.com/immich-app/immich/pull/1978
* chore(docs): Document fallback timezone setting (fixes #2000) by @etnoy in https://github.com/immich-app/immich/pull/2003
* chore(docs): Document cli path parameter change by @etnoy in https://github.com/immich-app/immich/pull/2011
* chore: Move away from docker hub where possible by @bo0tzz in https://github.com/immich-app/immich/pull/2006

## New Contributors
* @brighteyed made their first contribution in https://github.com/immich-app/immich/pull/1919
* @DragosRotaru made their first contribution in https://github.com/immich-app/immich/pull/1974
* @raisinbear made their first contribution in https://github.com/immich-app/immich/pull/1980
* @twitsforbrains made their first contribution in https://github.com/immich-app/immich/pull/1978
* @etnoy made their first contribution in https://github.com/immich-app/immich/pull/2002
* @truedem0n made their first contribution in https://github.com/immich-app/immich/pull/1997

**Full Changelog**: https://github.com/immich-app/immich/compare/v1.50.1...a

<hr /><em>This discussion was created from the release <a href='https://github.com/immich-app/immich/releases/tag/v1.51.0'>v1.51.0</a>.</em>

**Comment by jackbrown1993** [comment]
Great release, good job to all involved. I think we should update the release notes to clarify the some-random-text API key for Typesense.

**Reply by Stoobiedoo** [reply]
I'm getting this on start up:
WARNING: The my-random-text variable is not set. Defaulting to a blank string.

And this error: "connect ECONNREFUSED" on the webpage

I've added the strings to the docker-compose.yml as
      - TYPESENSE_API_KEY=${my-random-text}
And in the .env I added this to the end
TYPESENSE_API_KEY=my-random-text

Thanks for any help understanding how I messed it up.

**Reply by jrasm91** [reply]
`${my-random-text}` means it is looking for a env variable named `my-random-text`. You probably want `TYPESENSE_API_KEY=${TYPESENSE_API_KEY}`, which is how it is setup in the default docker-compose file.

**Reply by Stoobiedoo** [reply]
Thank you so much for the clarification. All fixed up now!

**Reply by elgatho** [reply]
Hello, when new version will be in App store? Just can't wait:)

**Reply by alextran1502** [reply]
> Hello, when new version will be in App store? Just can't wait:)

They are being reviewed by the app stores now

**Comment by MozPri** [comment]
I'm getting error 500 after the update.

`500 getaddrinfo ENOTFOUND immich-server` 

I'm using Portainer with the latest compose and .env files.
Also, the immich_typesense container does not start. State == exited.

Any suggestions?

**Reply by alextran1502** [reply]
Can you provide the server log?

**Reply by MozPri** [reply]
[_immich_server_logs.txt](https://github.com/immich-app/immich/files/11027793/_immich_server_logs.txt)

**Reply by alextran1502** [reply]
You might need to open an issue and provide your whole setup info so we can help. Make sure you bring all containers down then up again for a proper restart

**Reply by uninvitedguest** [reply]
@MozPri I'm running in portainer and received the same web error, but my issue was that while the typesense container would run, the server container would start and then immediately quit.

To resolve this I added the `env_file:` and `stack.env` lines to the compose, and I also removed any special characters from my API key (kept it at just letters and numbers). If you are using any custom networks, don't forget to add the typesense container to the custom network.

```
  typesense:
    container_name: immich_typesense
    image: typesense/typesense:0.24.0
    environment:
      - TYPESENSE_API_KEY=${TYPESENSE_API_KEY}
      - TYPESENSE_DATA_DIR=/data
    logging:
      driver: none
    volumes:
      - /mnt/ssd/immich/tsdata:/data
    env_file:
      - stack.env
    networks:
      - immich-internal
```

Note that I bound /data to a /tsdata volume on a separate drive (not that I know what goes in there). Nothing you need to replicate.

**Comment by uninvitedguest** [comment]
Thanks for your hard work on this! I'm exciting for the new functionality.

I got lucky that I took a deeper look at the .yaml before updating and caught that there was a new container. The [docker compose](https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml) file still notes Version 3.8, as it has for a long time. Would these changes not have warranted a version update to flag that there were changes to the .yaml?

**Reply by jrasm91** [reply]
The version key in the yaml file is the docker yaml specification version, not the immich server version.

See https://docs.docker.com/compose/compose-file/04-version-and-name/

**Reply by uninvitedguest** [reply]
Right, that's what I'm referring to. The .yaml received some updates (new container added, image references from docker hub changed to ghcr.io.)

Changing the yaml version would flag that there are changes to configuration that need to be captured when going through an update.

**Reply by alextran1502** [reply]
@uninvitedguest that is related to docker and docker-compose so it is not for Immich to use to keep track of the docker-compose.yml file changes

**Comment by Pixelartist** [comment]
I think this is great! Sadly it killed my setup since Typesense does not currently support an armv7 version:
`Failed to deploy a stack: typesense Pulling no matching manifest for linux/arm/v7 in the manifest list entries`
From the releases I can see they supported v8 a couple of times. Maybe you can see a quick way of fixing it?

**Reply by alextran1502** [reply]
The only way to fix this is for you to build the Typesense image on your machine and use that image in the compose file

**Reply by Pixelartist** [reply]
I looked through their github issues and it seems that currently there is no way to properly build it. [The issue is mentioning a request](https://github.com/typesense/typesense/issues/185) from a user trying to build it on a mac (the new ones with the arm chip) and no one yet has succeeded.
Since I am running it on a NAS, options are rather limited.

**Reply by bacmanni** [reply]
@Pixelartist Did you find a working solution for this? I am trying to get Immich working on my raspberry pi 4 and I am getting basically the same error with version 1.71.0. I found a one tutorial on how to get immich working on pi but that one uses older version. So I am kind of stuck without the info if immich is supposed to be even working on pi.

**Reply by raisinbear** [reply]
@bacmanni It works on a RaspberryPi, especially 4. You should, however, have the 64 bit version of raspbian if you want to easily deploy with typesense. They don’t have a 32 bit (armv7) build in their official docker repo afaik.

Edit: you can disable search and machine learning and comment the respective sections in the compose file. I see armv7 containers are still built for 1.71.0, so it Immich might still run on 32 bit. Haven’t checked myself though. typesense was an addition sometime this year I think. That would explain why old tutorials wouldn’t mention this particular problem.

**Reply by bacmanni** [reply]
@raisinbear Ah, so the problems is the 32bit OS. At the time of installing my Raspberry there were some stability issues with it (or at least that what I read) so I decided to with 32bit. Maybe it is time to give that another change. It might be more future proof solution generally. However, thanks for pointing me to a right direction.

**Comment by steve-human** [comment]
Shouldn't this also be included in `https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml`?  as well as `https://github.com/immich-app/immich/releases/latest/download/example.env`

As a new user I was having issues installing because TYPESENSE_API_KEY was required, but the documentation doesn't mention anything about typesense

**Reply by alextran1502** [reply]
This is release 1.52, we are on release 1.100 now and Typesense is no longer used

**Reply by steve-human** [reply]
Ugh, sorry about that. I must have a very old image on my system. Docker compose pull fixes the issue
